dali/python/nvidia/dali/plugin/base_iterator.py

```diff
@@ -0,0 +1,185 @@
+        """
+        if self._counter >= self._size and self._size > 0:
+            if self._auto_reset:
+                self.reset()
+            raise StopIteration
+
+    def _remove_padded(self):
+        """
+        Checks if remove any padded sample and how much
+        """
+        if_drop = False
+        left = -1
+        if not self._fill_last_batch:
+            # calculate each shard size for each id, and check how many samples are left by substracting
+            # from iterator counter the shard size, then go though all GPUs and check how much data needs to be dropped
+            shards_beg = np.floor(self._shards_id * self._size_no_pad / self._shards_num).astype(np.int)
+            shards_end = np.floor((self._shards_id + 1) * self._size_no_pad / self._shards_num).astype(np.int)
+            left = self.batch_size - (self._counter - (shards_end - shards_beg))
+            if_drop = np.less(left, self.batch_size)
+        return if_drop, left
+
+    def reset(self):
+        """
+        Resets the iterator after the full epoch.
+        DALI iterators do not support resetting before the end of the epoch
+        and will ignore such request.
+        """
+        if self._counter >= self._size or self._size < 0:
+            if self._fill_last_batch and not self._last_batch_padded:
+                if self._reader_name:
+                    # accurate way
+                    # get the number of samples read in this epoch by each GPU
+                    # self._counter had initial value of min(self._counter_per_gpu) so substract this to get the actual value
+                    self._counter -= min(self._counter_per_gpu)
+                    self._counter_per_gpu = self._counter_per_gpu + self._counter
+                    # check how much each GPU read ahead from next shard, as shards have different size each epoch
+                    # GPU may read ahead or not
+                    self._counter_per_gpu = self._counter_per_gpu - self._shard_sizes
+                    # to make sure that in the next epoch we read the whole shard we need to set start value to the smallest one
+                    self._counter = min(self._counter_per_gpu)
+                else:
+                    # legacy way
+                    self._counter = self._counter % self._size
+            else:
+                self._counter = 0
+            # advance to the next shard
+            if self._reader_name:
+                if not self._is_stick_to_shard:
+                    # move shards id for wrapped pipeliens
+                    self._shards_id = (self._shards_id + 1) % self._shards_num
+                # revaluate _size
+                if self._fill_last_batch and not self._last_batch_padded:
+                    # move all shards ids GPU ahead
+                    if not self._is_stick_to_shard:
+                        self._shard_sizes = np.roll(self._shard_sizes, 1)
```

Comment:
Is there any guarantee, that we didn't start with shard ids in some random order?
For example we got list of pipelines representing shards: [3, 2, 0, 1], after advancing to next shards, we would get [0, 3, 1, 2], which I guess for sizes looks like some random permutation and not a roll operation.

Reply:
`self._shard_sizes` are calculated in line `90` for all pipelines (0-num_shards). So we know the order is maintained.

Reply:
`self._shards_id` keeps only relevant ids - like [5, 3, 2, 10], but they are increased, not rolled.

Reply:
Ok, so why we even roll the shard_sizes? Shouldn't shard stay the same size for the whole time?

Reply:
We need to do that to know how much data through iterations every pipeline has read.